src/Microsoft.AspNetCore.ResponseCaching/Internal/ResponseCachingKeyProvider.cs

```diff
@@ -141,19 +144,26 @@ public string CreateStorageVaryByKey(ResponseCachingContext context)
                     if (varyByRules.QueryKeys.Count == 1 && string.Equals(varyByRules.QueryKeys[0], "*", StringComparison.Ordinal))
                     {
                         // Vary by all available query keys
-                        foreach (var query in context.HttpContext.Request.Query.OrderBy(q => q.Key, StringComparer.OrdinalIgnoreCase))
+                        var queryArray = context.HttpContext.Request.Query.ToArray();
+                        Array.Sort(queryArray, QueryKeyComparer.OrdinalIgnoreCase);
```

Comment:
Comment this one is ignore case because Request.Query aggregates keys using ignore case wherease the values are compared ordinally.